### PR TITLE
snapstate: support for pre-refresh hook

### PR DESCRIPTION
**Pull request by stolowski**
This branch completes the support for refresh hooks by providing pre-refresh hook that is run during refresh before switching to the new revision (as opposed to post-refresh hook).

Note: Depends on #4259.

### CODE REVIEW

File: overlord/hookstate/hooks.go [hunk 1]
```diff
@@ -53,6 +54,17 @@ func SetupPostRefreshHook(st *state.State, snapName string) *state.Task {
 	}
 
 	summary := fmt.Sprintf(i18n.G("Run post-refresh hook of %q snap if present"), hooksup.Snap)
+	return HookTask(st, summary, hooksup, nil)
+}
+
+func SetupPreRefreshHook(st *state.State, snapName string) *state.Task {
```
Comment (zyga) [inline overlord/hookstate/hooks.go]: As a general comment it'd be nice if we had a no-op thing here when the snap in question doesn't have a particular hook.

File: overlord/snapstate/snapstate.go [hunk 2]
```diff
@@ -134,7 +137,15 @@ func doInstall(st *state.State, snapst *SnapState, snapsup *SnapSetup, flags int
 		stop := st.NewTask("stop-snap-services", fmt.Sprintf(i18n.G("Stop snap %q services"), snapsup.Name()))
 		addTask(stop)
 		prev = stop
+	}
+
+	if runRefreshHooks {
+		preRefreshHook := SetupPreRefreshHook(st, snapsup.Name())
```
Comment (zyga) [inline overlord/snapstate/snapstate.go]: Can the refresh hook restart the services because it runs after they are stopped and can poke their sockets and what not?

File: overlord/state/taskrunner.go [hunk 3]
```diff
@@ -336,7 +336,15 @@ func (r *TaskRunner) Ensure() {
 			}
 			continue
 		}
+
+		if mustWait(t) {
```
Comment (zyga) [inline overlord/state/taskrunner.go]: This is from another PR, right?

Comment (stolowski) [reply overlord/state/taskrunner.go]: Right.

File: overlord/snapstate/snapstate.go (hunk 2 above)
Comment (stolowski) [reply overlord/snapstate/snapstate.go]: Yes, it can, there is nothing that prevents it currently. I see how this can be an issue that leads to a failure if service is started for the old revision of the snap in pre-refresh. We can easily prevent this and error out in snapctl based on hook name, if that's what we want.

File: overlord/hookstate/hooks.go (hunk 1 above)
Comment (stolowski) [reply overlord/hookstate/hooks.go]: That applies to the hook machinery, although IMHO it's useful that we have the task logged and visible even if there is no hook ("Run foo hook of bar snap if present" - that's what we see in the tasks). In any case it's a material for separate discussion if we want that.

**Comment by zyga** [comment]
I'm marking this as blocked as the prerequisite PR is still in a state where changes are requested.

File: overlord/hookstate/hooks.go (hunk 1 above)
Comment (pedronis) [reply overlord/hookstate/hooks.go]: in general we might now even know if the hook is there (because of implicit hooks) when we setup the tasks in a change

File: overlord/snapstate/snapstate.go (hunk 2 above)
Comment (niemeyer) [reply overlord/snapstate/snapstate.go]: Why are we running this after the services are stopped?  Part of the reason of running this before it stops is precisely so that we have the data and the old environment at hand. People will need the service running if we intend them to be able to dump data before the refresh, for example.

File: overlord/snapstate/snapstate.go [hunk 4]
```diff
@@ -134,7 +137,15 @@ func doInstall(st *state.State, snapst *SnapState, snapsup *SnapSetup, flags int
...
+		addTask(preRefreshHook)
+		prev = preRefreshHook
+	}
 
+	if snapst.Active {
```
Comment (niemeyer) [inline overlord/snapstate/snapstate.go]: Also, this sounds suspect. Does it mean we're running the hook even when the snap is not active? In which context would this make sense, or otherwise in which ways will it create problems?

Comment (stolowski) [reply overlord/snapstate/snapstate.go]: You're right about running the hook before stopping rather than after. I missed this aspect as I considered pre- and post- refresh hooks to be about running before and after switching active revision.

**Comment by stolowski** [comment]
Fixed per latest agreements.

File: overlord/snapstate/snapstate.go [hunk 5]
```diff
@@ -61,6 +61,10 @@ func needsMaybeCore(typ snap.Type) int {
 }
 
 func doInstall(st *state.State, snapst *SnapState, snapsup *SnapSetup, flags int) (*state.TaskSet, error) {
+	if snapst.IsInstalled() && !snapst.Active {
+		return nil, fmt.Errorf("internal error: cannot update a disabled snap")
```
Comment (niemeyer) [inline overlord/snapstate/snapstate.go]: There's no reason for this to be an internal error, as this is a real situation that can happen by the user manipulating the snap state in sane ways. We just want to make sure that case is forbidden here.

File: overlord/snapstate/snapstate.go [hunk 6]
```diff
@@ -122,14 +126,23 @@ func doInstall(st *state.State, snapst *SnapState, snapsup *SnapSetup, flags int
 		prev = checkAsserts
 	}
 
+	// run refresh hooks when updating existing snap, otherwise run install
+	runRefreshHooks := (snapst.IsInstalled() && !snapsup.Flags.Revert)
```
Comment (niemeyer) [inline overlord/snapstate/snapstate.go]: This should be moved right above the condition that uses it below. Otherwise this is lost between code that is completely unrelated to it.

File: overlord/snapstate/snapstate.go [hunk 7]
```diff
@@ -122,14 +126,23 @@ func doInstall(st *state.State, snapst *SnapState, snapsup *SnapSetup, flags int
 		prev = checkAsserts
 	}
 
+	runRefreshHooks := (snapst.IsInstalled() && !snapsup.Flags.Revert)
```
Comment (pedronis) [inline overlord/snapstate/snapstate.go]: I think Gustavo meant the flag should also be computed near first use

File: overlord/snapstate/snapstate.go [hunk 8]
```diff
@@ -61,6 +61,10 @@ func needsMaybeCore(typ snap.Type) int {
 }
 
 func doInstall(st *state.State, snapst *SnapState, snapsup *SnapSetup, flags int) (*state.TaskSet, error) {
+	if snapst.IsInstalled() && !snapst.Active {
+		return nil, fmt.Errorf("cannot update a disabled snap %q", snapsup.Name())
```
Comment (pedronis) [inline overlord/snapstate/snapstate.go]: article "a" is not needed if we have the name
